doc/design/field_evaluate.md

```diff
@@ -0,0 +1,161 @@
+- reference to the EvalPoints instance (therefore it should not be dim templated either)
+- dimension (probably just for checks)
+- mask which points (of the EvalPoints) are used in passed `EvalSubset` objects, only values in these points has to be evaluated.
+- allocation informations are just collected, actual allocation is performed during first update call
+  alternatively we can introduce explicit function call to do that.
+
+**Example**
+```
+    // still in class Assembly<dim> definition
+   this->mass_fields.cache_allocate(this->mass_eval, this->mapping);
+   this->side_fields.cache_allocate(this->side_eval, this->mapping);
+} // end of Assembly<dim>
+```
+TODO:
+- we must use lazy cache allocation since the information is added to the fields in different allocation calls and with different dimensions. 
+
+### 5.2 Map element to the cache element index
+```
+void Assembly::mass_assembly(DHCellAccessor cell) {
+    // this can possibly be called also in the generic assembly loop as 
+    // the cache prefetching
+    // assume that ngh_el, is a neigbour element
+    
+    // map used element into element cache
+    el_cache = this->element_cache_map(cell.element());
+    ngh_el_cache = this->element_cache_map(ngh_el.element());
+
+```
+### 5.1 (and 4.) Cache Update
+```
+    // just call cache_update for the fields in the field sets
+    this->mass_fields.cache_update(this->element_cache_map);
+    this->face_fields.cache_update(this->element_cache_map);
+```
+### 5.3 Cache read
+```
+    // Bulk integral, no sides, no permutations.
+    for(auto q_point: this->mass_eval.points()) {
+        // Extracting the cached values.
+        double cs = cross_section.value(el_cache, q_point);
+        double ngh_cs = cross_section.value(ngh_el_cache, q_point);
+        // This would be nice to have. Not clear how to 
+        // deal with more then single element as fe_values have its own cache that has to be updated.
+        double base_fn = presssure_field_fe.fe_values.value(q_point);
+    }   
+
+    // Side integrals.
+    for (auto side : cell.side_range()) {
+        double side_avg = 0;
+        // vector of local side quadrature points in the correct side permutation
+	auto side_points = this->side_eval.points(side->side())
+	for (auto p : side_points)
+            side_avg += cross_section.value(el_cache, p) * sigma.value(el_cache, p);
+    }
+}
+```
+- `Field<spacedim, Value>::value(EvalElement eval_el, EvalPoint q_point)`
+  returns Value, i.e. some FieldValue<..>. The implementation should do something like:
+  ```return this->value_cache[eval_el.dim()].value(eval_el.cache_idx(), q_point.cache_idx())```
+  Where `value_cache` is the instance of `FieldValueCache`.
+- `this->mass_eval` and `this->side_eval` are instances of `EvalSubset`.
+
+
+
+
+**Further thoughts**
+- Extension to interdependent fields: If a field depends on the other field, it recursively informs the other field about quadrature.
+
+
+
```

Comment:
There is no class defining PointAccessors in desing.  I am not sure if the class is missing in the design or it is intention and we don't need accessor class.

Reply:
EvalPoint is used instead of accessor. User need not to have access to the actual point so we need any mean to pass identification of the local point so it is in fact just an index.